### DISCUSSION TITLE
How to create scala-cli type of "lightweight launcher JAR" from a project?

### DISCUSSION THREAD

**Discussion by jk-1** [com-lihaoyi/mill#2647] [Q&A]
How can I create a "lightweight launcher JAR" from my project with `mill`? 
I.e. How to create a single executable file same way as `scala-cli package mySource.scala` produces?

See default package format: https://scala-cli.virtuslab.org/docs/commands/package

**Comment by lihaoyi** [comment]
Probably `./mill foo.assembly`?

**Comment by jk-1** [comment]
I think they are not same thing.

The mill assembly command creates fat jar that I can run offline 
while scala-cli package creates smaller executable file which downloads missing pieces if they do not exist.

> The default package format writes a lightweight launcher JAR, like the "bootstrap" JAR files [generated by coursier](https://get-coursier.io/docs/cli-bootstrap#bootstraps). These JARs tend to have a small size (mostly containing only the byte code from your own sources), can be generated fast, and download their dependencies upon first launch via [coursier](https://get-coursier.io/).


See below a small example: HelloScript.sc

```
val sv = scala.util.Properties.versionNumberString

val message = s"Hello from Scala ${sv}, Java ${System.getProperty("java.version")}"
println(message)
println("args: " + args.toSeq)

```

```
ll out/project/assembly.dest/out*
-rwx--x--x 1 jk jk 7300804 heinä   1 15:52 out/project/assembly.dest/out.jar*

ll HelloScript
-rwx------ 1 jk jk 171427 heinä   1 15:54 HelloScript*

```

**Comment by lihaoyi** [comment]
Oh you want a dependency-downloading bootstrapper. Mill does not have that built in. You should be able to implement it as a custom task in your own build I think
